File: next-frontend/api/definitions/models/notification.ts
import { CategoryWithIdSchema } from "@/api/definitions/models/category";
import {
  ReadFeedReactionSchema,
  ReadUserAvatarSchema,
} from "@/api/definitions/models/feed";
import { z } from "zod";
import { UserSchema } from "@/api/definitions/models/user";

export const NotificationSourceTypeSchema = z.enum(["user", "system"]);

export const NotificationTypeSchema = z.enum([
  "friend:new_request",
  "friend:request_accepted",
  "session:reaction_added",
  "system:new_release",
  "admin:sandbox:failed-deploy",
  "admin:backup:completed",
  "admin:backup:failed",
]);

export const FriendRequestDataSchema = z.object({
  message: z.string().nullish(),
  request_id: z.string(),
  requestor: ReadUserAvatarSchema,
});

export const FriendRequestAcceptedDataSchema = z.object({
  accepter: ReadUserAvatarSchema,
});

export const SessionReactionDataSchema = z
  .object({
    session_category: CategoryWithIdSchema,
    session_end_time: z.coerce.date<Date>(),
    session_id: z.string(),

    session_start_time: z.coerce.date<Date>(),
  })
  .and(ReadFeedReactionSchema);

export const SystemReleaseDataSchema = z.object({
  release_id: z.string(),
  short_description: z.string().optional(),
  title: z.string(),
});

export const SystemNotificationDataSchema = z.object({
  system_id: z.string(),
  system_name: z.string(),
});

export const SandboxFailedDeployDataSchema = z.object({
  sandbox_lifecycle_id: z.string().uuid(),
});

export const BackupCompletedDataSchema = z.object({
  backup_id: z.number(),
  backup_size_bytes: z.number().nullable(),
  duration_seconds: z.number().nullable(),
  started_at: z.coerce.date().nullable(),
});

export const BackupFailedDataSchema = z.object({
  backup_id: z.number(),
  duration_seconds: z.number().nullable(),
  error_message: z.string().nullable(),
  started_at: z.coerce.date().nullable(),
});

export const NotificationSourceSchema = z.discriminatedUnion("source_type", [
  z.object({
    source_data: UserSchema,
    source_type: z.literal("user"),
  }),
  z.object({
    source_data: SystemNotificationDataSchema,
    source_type: z.literal("system"),
  }),
]);

export const NotificationDataSchema = z.discriminatedUnion(
  "notification_type",
  [
    z.object({
      data: FriendRequestDataSchema,
      notification_type: z.literal("friend:new_request"),
    }),
    z.object({
      data: FriendRequestAcceptedDataSchema,
      notification_type: z.literal("friend:request_accepted"),
    }),
    z.object({
      data: SessionReactionDataSchema,
      notification_type: z.literal("session:reaction_added"),
    }),
    z.object({
      data: SystemReleaseDataSchema,
      notification_type: z.literal("system:new_release"),
    }),
    z.object({
      data: SandboxFailedDeployDataSchema,
      notification_type: z.literal("admin:sandbox:failed-deploy"),
    }),
    z.object({
      data: BackupCompletedDataSchema,
      notification_type: z.literal("admin:backup:completed"),
    }),
    z.object({
      data: BackupFailedDataSchema,
      notification_type: z.literal("admin:backup:failed"),
    }),
  ],
);

export const NotificationSchema = z
  .object({
    created_at: z.coerce.date<Date>(),
    id: z.string(),
    seen: z.boolean(),
    user_id: z.string(),
  })
  .and(NotificationSourceSchema)
  .and(NotificationDataSchema);

export type BackupCompletedData = z.infer<typeof BackupCompletedDataSchema>;
export type BackupFailedData = z.infer<typeof BackupFailedDataSchema>;
export type FriendRequestAcceptedData = z.infer<
  typeof FriendRequestAcceptedDataSchema
>;
export type NewFriendRequestData = z.infer<typeof FriendRequestDataSchema>;
export type NewReleaseData = z.infer<typeof SystemReleaseDataSchema>;

export type Notification = z.infer<typeof NotificationSchema>;
export type NotificationData = z.infer<typeof NotificationDataSchema>;
export type NotificationSource = z.infer<typeof NotificationSourceSchema>;
export type NotificationSourceType = z.infer<
  typeof NotificationSourceTypeSchema
>;
export type NotificationType = z.infer<typeof NotificationTypeSchema>;
export type SandboxFailedDeployData = z.infer<
  typeof SandboxFailedDeployDataSchema
>;
export type SessionReactionAddedData = z.infer<
  typeof SessionReactionDataSchema
>;
export type SystemNotificationData = z.infer<
  typeof SystemNotificationDataSchema
>;

File: backend/src/entity/notification.rs
use chrono::{DateTime, Local};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::{
    dto::{feed::ReadFeedReactionDto, user::read_user::ReadUserDto},
    entity::category::Category,
};

#[derive(Clone, Debug, PartialEq, PartialOrd, sqlx::Type, Deserialize, Serialize)]
#[sqlx(type_name = "notification_type")]
#[serde(rename_all = "kebab-case")]
pub enum NotificationTypeSql {
    #[sqlx(rename = "friend:new_request")]
    #[serde(rename = "friend:new_request")]
    FriendNewRequest,

    #[sqlx(rename = "friend:request_accepted")]
    #[serde(rename = "friend:request_accepted")]
    FriendRequestAccepted,

    #[sqlx(rename = "session:reaction_added")]
    #[serde(rename = "session:reaction_added")]
    SessionReactionAdded,

    #[sqlx(rename = "system:new_release")]
    #[serde(rename = "system:new_release")]
    SystemNewRelease,

    #[sqlx(rename = "task:completed")]
    #[serde(rename = "task:completed")]
    TaskCompleted,

    #[sqlx(rename = "project:completed")]
    #[serde(rename = "project:completed")]
    ProjectCompleted,
}

#[derive(Clone, Debug, PartialEq, PartialOrd, sqlx::Type, Deserialize, Serialize)]
#[sqlx(type_name = "notification_source_type")]
#[serde(rename_all = "lowercase")]
pub enum NotificationSourceTypeSql {
    #[sqlx(rename = "user")]
    User,

    #[sqlx(rename = "system")]
    System,
}

#[derive(Clone, Serialize, Deserialize, Debug)]
pub struct FriendRequestData {
    pub requestor: ReadUserDto,
    pub message: Option<String>,
    pub request_id: Uuid,
}

#[derive(Clone, Serialize, Deserialize, Debug)]
pub struct FriendRequestAcceptedData {
    pub accepter: ReadUserDto,
}

#[derive(Clone, Serialize, Deserialize, Debug)]
pub struct SessionReactionData {
    #[serde(flatten)]
    pub reaction: ReadFeedReactionDto,
    pub session_id: Uuid,
    pub session_category: Category,
    pub session_start_time: DateTime<Local>,
    pub session_end_time: DateTime<Local>,
}

#[derive(Clone, Serialize, Deserialize, Debug)]
pub struct SystemReleaseData {
    pub release_id: Uuid,
    pub title: String,
    pub short_description: Option<String>,
}

#[derive(Clone, Serialize, Deserialize, Debug)]
pub struct TaskCompletedData {
    pub task_id: Uuid,
    pub task_name: String,
    pub project_name: String,
    pub total_hours: f64,
}

#[derive(Clone, Serialize, Deserialize, Debug)]
pub struct ProjectCompletedData {
    pub project_id: Uuid,
    pub project_name: String,
    pub total_tasks: i64,
    pub total_hours: f64,
}

#[derive(Clone, Serialize, Deserialize, Debug)]
#[serde(tag = "notification_type", content = "data", rename_all = "kebab-case")]
pub enum NotificationType {
    #[serde(rename = "friend:new_request")]
    FriendNewRequest(FriendRequestData),

    #[serde(rename = "friend:request_accepted")]
    FriendRequestAccepted(FriendRequestAcceptedData),

    #[serde(rename = "session:reaction_added")]
    SessionReactionAdded(SessionReactionData),

    #[serde(rename = "system:new_release")]
    SystemNewRelease(SystemReleaseData),

    #[serde(rename = "task:completed")]
    TaskCompleted(TaskCompletedData),

    #[serde(rename = "project:completed")]
    ProjectCompleted(ProjectCompletedData),
}

#[derive(Clone, Serialize, Deserialize, Debug)]
pub struct SystemNotificationData {
    pub system_id: String,
    pub system_name: String,
}

#[derive(Clone, Serialize, Deserialize, Debug)]
#[serde(tag = "source_type", content = "source_data", rename_all = "lowercase")]
pub enum NotificationSource {
    User(ReadUserDto),
    System(SystemNotificationData),
}

#[derive(Clone, Serialize, Deserialize, Debug)]
pub struct Notification {
    pub id: Uuid,
    pub user_id: String,
    #[serde(flatten)]
    pub source: NotificationSource,
    #[serde(flatten)]
    pub notification_type: NotificationType,
    pub seen: bool,
    pub created_at: DateTime<Local>,
}
